In [7]:
import pandas as pd

df = pd.read_csv('ecommerce_orders_dirty.csv')
df

Order ID Customer Name                   Email         Country Currency  \
0      ORD8817        😊 Alex         😊alex@gmail.con              DE      USD   
1      ORD2676   Marie Curie  mariecurie@hotmail.com             USA        $   
2      ORD5686        😊 Alex       😊alex@outlook.com  United Kingdom        $   
3     ORD19599    Jane Smith   janesmith@outlook.com     Deutschland      USD   
4      ORD6512   Marie Curie  mariecurie@hotmail.com              UK        €   
...        ...           ...                     ...             ...      ...   
9995   ORD3336    Jane Smith   janesmith@hotmail.com              UK      USD   
9996  ORD15481   Anna Müller  annamüller@hotmail.com         Germany        $   
9997   ORD2267            王伟            王伟@gmail.con              UK      JPY   
9998  ORD14299            王伟          王伟@hotmail.com              DE        $   
9999   ORD8011   Carlos Diaz  carlosdiaz@outlook.com     Deutschland      USD   

     Product SKU    Product Description Quantity Unit Price  Total Price  \
0        SKU-123             Smartphone        2     twenty          NaN   
1            345             Smartphone        1         40          NaN   
2        #SKU234        Laptop&nbsp;Bag        0      15,00          NaN   
3        SKU-123            USB-C Cable    three        €30          NaN   
4            345           Gaming Chair       -1      15,00          NaN   
...          ...                    ...      ...        ...          ...   
9995     #SKU234            USB-C Cable    three     twenty          NaN   
9996     SKU-123            USB-C Cable        2     twenty          NaN   
9997     SKU_999            USB-C Cable        2     twenty          NaN   
9998     #SKU234           Gaming Chair        1         25          NaN   
9999     sku-678  <b>Wireless Mouse</b>        0         25          NaN   

      Order Date Shipping Fee Delivery Status Delivery Date Payment Method  \
0     2026-12-01         5.00       Delivered    2026-12-11    Credit Curd   
1     2023-05-12            7         delivrd    2023-05-17         paypal   
2     2023-05-12         free         delivrd    2023-05-17    credit card   
3     2026-12-01           €4       Delivered    2026-12-03         paypal   
4     2026-12-01           $5       Cancelled    2026-11-30  Cash on deliv   
...          ...          ...             ...           ...            ...   
9995  2023-05-12            7         delivrd    2023-05-11    credit card   
9996  2023-05-12            7       Delivered    2023-05-17    Credit Curd   
9997  2026-12-01           €4         delivrd    2026-12-11    credit card   
9998  2026-12-01            7       Cancelled    2026-12-06  Cash on deliv   
9999  2023-05-12           -3       Delivered    2023-05-14         paypal   

                 Notes  Unnamed: 0  
0          Excellent 👍           0  
1     <div>Hurry</div>           1  
2          Excellent 👍           2  
3          Excellent 👍           3  
4                  NaN           4  
...                ...         ...  
9995               NaN        9995  
9996       Excellent 👍        9996  
9997               NaN        9997  
9998               NaN        9998  
9999               NaN        9999  

[10000 rows x 17 columns]

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Order ID             8967 non-null   object 
 1   Customer Name        10000 non-null  object 
 2   Email                10000 non-null  object 
 3   Country              10000 non-null  object 
 4   Currency             10000 non-null  object 
 5   Product SKU          10000 non-null  object 
 6   Product Description  10000 non-null  object 
 7   Quantity             10000 non-null  object 
 8   Unit Price           10000 non-null  object 
 9   Total Price          0 non-null      float64
 10  Order Date           10000 non-null  object 
 11  Shipping Fee         10000 non-null  object 
 12  Delivery Status      10000 non-null  object 
 13  Delivery Date        10000 non-null  object 
 14  Payment Method       10000 non-null  object 
 15  Notes                5981 non-null   

# Inspect the data

In [6]:
df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Order ID             8967 non-null   object 
 1   Customer Name        10000 non-null  object 
 2   Email                10000 non-null  object 
 3   Country              10000 non-null  object 
 4   Currency             10000 non-null  object 
 5   Product SKU          10000 non-null  object 
 6   Product Description  10000 non-null  object 
 7   Quantity             10000 non-null  object 
 8   Unit Price           10000 non-null  object 
 9   Total Price          0 non-null      float64
 10  Order Date           10000 non-null  object 
 11  Shipping Fee         10000 non-null  object 
 12  Delivery Status      10000 non-null  object 
 13  Delivery Date        10000 non-null  object 
 14  Payment Method       10000 non-null  object 
 15  Notes                5981 non-null   

Order ID Customer Name           Email  Country Currency Product SKU  \
count      8967         10000           10000    10000    10000       10000   
unique     7191             7              28        7        5           5   
top     ORD6819            王伟  王伟@hotmail.com  Germany      USD     SKU-123   
freq          5          1457             382     1467     2064        2068   
mean        NaN           NaN             NaN      NaN      NaN         NaN   
std         NaN           NaN             NaN      NaN      NaN         NaN   
min         NaN           NaN             NaN      NaN      NaN         NaN   
25%         NaN           NaN             NaN      NaN      NaN         NaN   
50%         NaN           NaN             NaN      NaN      NaN         NaN   
75%         NaN           NaN             NaN      NaN      NaN         NaN   
max         NaN           NaN             NaN      NaN      NaN         NaN   

       Product Description Quantity Unit Price  Total Price  Order Date  \
count                10000    10000      10000          0.0       10000   
unique                   5        7          7          NaN           2   
top             Smartphone        3      £12.5          NaN  2023-05-12   
freq                  2051     1479       1479          NaN        7456   
mean                   NaN      NaN        NaN          NaN         NaN   
std                    NaN      NaN        NaN          NaN         NaN   
min                    NaN      NaN        NaN          NaN         NaN   
25%                    NaN      NaN        NaN          NaN         NaN   
50%                    NaN      NaN        NaN          NaN         NaN   
75%                    NaN      NaN        NaN          NaN         NaN   
max                    NaN      NaN        NaN          NaN         NaN   

       Shipping Fee Delivery Status Delivery Date Payment Method        Notes  \
count         10000           10000         10000          10000         5981   
unique            6               5             8              5            3   
top               7         Shipped    2023-05-22    credit card  Excellent 👍   
freq           1686            2079          1895           2099         2028   
mean            NaN             NaN           NaN            NaN          NaN   
std             NaN             NaN           NaN            NaN          NaN   
min             NaN             NaN           NaN            NaN          NaN   
25%             NaN             NaN           NaN            NaN          NaN   
50%             NaN             NaN           NaN            NaN          NaN   
75%             NaN             NaN           NaN            NaN          NaN   
max             NaN             NaN           NaN            NaN          NaN   

         Unnamed: 0  
count   10000.00000  
unique          NaN  
top             NaN  
freq            NaN  
mean     4999.50000  
std      2886.89568  
min         0.00000  
25%      2499.75000  
50%      4999.50000  
75%      7499.25000  
max      9999.00000

# Count nulls per column

In [7]:
nulls_per_column = df.isnull().sum()
print("Nulls per column:\n", nulls_per_column)

Nulls per column:
 Order ID                1033
Customer Name              0
Email                      0
Country                    0
Currency                   0
Product SKU                0
Product Description        0
Quantity                   0
Unit Price                 0
Total Price            10000
Order Date                 0
Shipping Fee               0
Delivery Status            0
Delivery Date              0
Payment Method             0
Notes                   4019
Unnamed: 0                 0
dtype: int64


# Check for duplicate Order IDs

In [8]:
duplicate_order_ids = df['Order ID'].duplicated().sum()
print(f"Number of duplicate Order IDs: {duplicate_order_ids}")

Number of duplicate Order IDs: 2808


# Check unique values in 'Currency' and 'Order Date' columns

In [10]:
print("Unique Currency formats:", df['Currency'].unique())
print("Unique Order Date formats:", df['Order Date'].unique())

Unique Currency formats: ['USD' '$' '€' 'JPY' '£']
Unique Order Date formats: ['2026-12-01' '2023-05-12']


# Check example patterns (first 20 unique values)

In [11]:
print(df['Order Date'].dropna().unique()[:20])

['2026-12-01' '2023-05-12']


#### Checkpoint: 3 Unique Data Issues


1. Several columns contain missing (null) values, especially in `Shipping Address` and `Order Date`.
2. There are duplicate entries in the `Order ID` column, which may indicate repeated or erroneous orders.
3. The `Order Date` and `Currency` columns contain inconsistent formats, making it difficult to analyze or convert them directly.

In [12]:
print(df.columns.tolist())

['Order ID', 'Customer Name', 'Email', 'Country', 'Currency', 'Product SKU', 'Product Description', 'Quantity', 'Unit Price', 'Total Price', 'Order Date', 'Shipping Fee', 'Delivery Status', 'Delivery Date', 'Payment Method', 'Notes', 'Unnamed: 0']


In [15]:
print(df.head())
print(df.sample(5))  # Shows 5 random rows# Shows the first 5 rows

   Order ID Customer Name                   Email         Country Currency  \
0   ORD8817        😊 Alex         😊alex@gmail.con              DE      USD   
1   ORD2676   Marie Curie  mariecurie@hotmail.com             USA        $   
2   ORD5686        😊 Alex       😊alex@outlook.com  United Kingdom        $   
3  ORD19599    Jane Smith   janesmith@outlook.com     Deutschland      USD   
4   ORD6512   Marie Curie  mariecurie@hotmail.com              UK        €   

  Product SKU Product Description Quantity Unit Price  Total Price  \
0     SKU-123          Smartphone        2     twenty          NaN   
1         345          Smartphone        1         40          NaN   
2     #SKU234     Laptop&nbsp;Bag        0      15,00          NaN   
3     SKU-123         USB-C Cable    three        €30          NaN   
4         345        Gaming Chair       -1      15,00          NaN   

   Order Date Shipping Fee Delivery Status Delivery Date Payment Method  \
0  2026-12-01         5.00       De

In [16]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Order ID             8967 non-null   object 
 1   Customer Name        10000 non-null  object 
 2   Email                10000 non-null  object 
 3   Country              10000 non-null  object 
 4   Currency             10000 non-null  object 
 5   Product SKU          10000 non-null  object 
 6   Product Description  10000 non-null  object 
 7   Quantity             10000 non-null  object 
 8   Unit Price           10000 non-null  object 
 9   Total Price          0 non-null      float64
 10  Order Date           10000 non-null  object 
 11  Shipping Fee         10000 non-null  object 
 12  Delivery Status      10000 non-null  object 
 13  Delivery Date        10000 non-null  object 
 14  Payment Method       10000 non-null  object 
 15  Notes                5981 non-null   

In [17]:
df.head(10).to_csv("sample_data.csv", index=False)

# Part 2: Advanced Cleaning Tasks
## Step A: Clean Structural Issues

### 1. Drop "Unnamed: 0" Column

In [19]:
df = df.drop(columns=['Unnamed: 0'])

### 2. Drop Rows with Missing "Order ID" or "Email"

In [20]:
df = df.dropna(subset=['Order ID', 'Email'])

### 3. Standardize "Customer Name" (title case, remove emojis/special characters)

In [21]:
import re

def clean_name(name):
    # Remove emojis and special characters except spaces and letters
    name = re.sub(r'[^\w\s]', '', name)
    name = re.sub(r'[\d_]', '', name)
    name = name.encode('ascii', 'ignore').decode('ascii')  # remove non-ASCII
    return name.title().strip()

df['Customer Name'] = df['Customer Name'].astype(str).apply(clean_name)

## Step B: Normalize Multi-format Fields
### 1. Convert "Order Date" to datetime and remove future dates

In [22]:
import pandas as pd
from datetime import datetime

# Convert Order Date to datetime
df['Order Date'] = pd.to_datetime(df['Order Date'], errors='coerce')

# Remove future dates (dates after today)
df = df[df['Order Date'] <= pd.Timestamp(datetime.now())]

### 2. Standardize "Currency" to "USD"


In [25]:
def currency_to_usd(row):
    currency = row['Currency']
    unit_price = str(row['Unit Price']).replace(',', '').replace(' ', '')
    try:
        unit_price = float(re.sub(r'[^\d.]+', '', unit_price))
    except:
        return None
    if currency == 'USD' or currency == '$':
        return unit_price
    elif currency == '€':
        return round(unit_price / 1.1, 2)
    elif currency == '£':
        return round(unit_price * 1.25, 2)
    elif currency == 'JPY':
        return round(unit_price * 0.007, 2)
    else:
        return unit_price  # If unknown, leave as is

df['Unit Price'] = df.apply(currency_to_usd, axis=1)
df['Currency'] = 'USD'

### 3. Clean and convert "Unit Price", "Shipping Fee", and "Total Price" to float

In [27]:
def clean_float(val):
    if pd.isnull(val):
        return 0.0
    val = str(val).replace(',', '').replace('free', '0').replace('Free', '0')
    val = re.sub(r'[^\d.-]+', '', val)
    try:
        return float(val)
    except:
        return 0.0

df['Unit Price'] = df['Unit Price'].apply(clean_float)
df['Shipping Fee'] = df['Shipping Fee'].apply(clean_float)
df['Total Price'] = df['Total Price'].apply(clean_float)

### 4. Clean "Quantity" (convert to numeric values)

In [17]:
def clean_quantity(val):
    if pd.isnull(val):
        return 0
    val_str = str(val).strip().lower()
    words_to_num = {'one': 1, 'two': 2, 'three': 3, 'four': 4, 'zero': 0, '-1': -1}
    if val_str in words_to_num:
        return words_to_num[val_str]
    try:
        return int(val_str)
    except:
        return 0
df = df.copy()
df['Quantity'] = df['Quantity'].apply(clean_quantity)

### 5. Recalculate "Total Price"

In [29]:
df['Total Price'] = df['Quantity'] * df['Unit Price'] + df['Shipping Fee']

In [31]:
df.describe(include='all')

Order ID Customer Name         Email  Country Currency Product SKU  \
count       6699          6699          6699     6699     6699        6699   
unique      5662             7            28        7        1           5   
top     ORD11021                王伟@gmail.con  Germany      USD     SKU-123   
freq           5           985           260      983     6699        1401   
mean         NaN           NaN           NaN      NaN      NaN         NaN   
min          NaN           NaN           NaN      NaN      NaN         NaN   
25%          NaN           NaN           NaN      NaN      NaN         NaN   
50%          NaN           NaN           NaN      NaN      NaN         NaN   
75%          NaN           NaN           NaN      NaN      NaN         NaN   
max          NaN           NaN           NaN      NaN      NaN         NaN   
std          NaN           NaN           NaN      NaN      NaN         NaN   

       Product Description     Quantity   Unit Price  Total Price  \
count                 6699  6699.000000  6699.000000  6699.000000   
unique                   5          NaN          NaN          NaN   
top             Smartphone          NaN          NaN          NaN   
freq                  1371          NaN          NaN          NaN   
mean                   NaN     1.145992   191.096286   225.282097   
min                    NaN    -1.000000     0.000000 -1878.000000   
25%                    NaN     0.000000     0.210000     0.000000   
50%                    NaN     1.000000    12.500000     5.540000   
75%                    NaN     3.000000    31.250000    39.080000   
max                    NaN     3.000000  1875.000000  5632.000000   
std                    NaN     1.461419   496.380528   978.237446   

                           Order Date  Shipping Fee Delivery Status  \
count                            6699   6699.000000            6699   
unique                            NaN           NaN               5   
top                               NaN           NaN       Delivered   
freq                              NaN           NaN            1404   
mean    2023-05-12 00:00:00.000000256      3.018062             NaN   
min               2023-05-12 00:00:00     -3.000000             NaN   
25%               2023-05-12 00:00:00      0.000000             NaN   
50%               2023-05-12 00:00:00      5.000000             NaN   
75%               2023-05-12 00:00:00      5.000000             NaN   
max               2023-05-12 00:00:00      7.000000             NaN   
std                               NaN      3.422146             NaN   

       Delivery Date Payment Method             Notes  
count           6699           6699              3983  
unique             4              5                 3  
top       2023-05-22    credit card  <div>Hurry</div>  
freq            1705           1419              1346  
mean             NaN            NaN               NaN  
min              NaN            NaN               NaN  
25%              NaN            NaN               NaN  
50%              NaN            NaN               NaN  
75%              NaN            NaN               NaN  
max              NaN            NaN               NaN  
std              NaN            NaN               NaN

In [32]:
print("Unique currencies:", df['Currency'].unique())
print("Order Date range:", df['Order Date'].min(), "to", df['Order Date'].max())

Unique currencies: ['USD']
Order Date range: 2023-05-12 00:00:00 to 2023-05-12 00:00:00


# Part 2: Advanced Cleaning Tasks (Continued)

## Step C: Fix Text & Codes

### 1. Email: Fix `.con` to `.com`, lowercase, drop invalid formats

In [10]:
import re

# Fix ".con" to ".com" and lowercase
df['Email'] = df['Email'].str.replace('.con', '.com', case=False).str.lower()

# Drop rows where email is not a valid format
df = df[df['Email'].str.match(r'^[\w\.-]+@[\w\.-]+\.\w+$')]

### 2. Product SKU: Extract numeric ID (e.g., "123" from "SKU-123" or "#SKU123")

In [18]:
def extract_sku(s):
    match = re.search(r'(\d+)', str(s))
    return match.group(1) if match else None

df['Product SKU'] = df['Product SKU'].apply(extract_sku)

### 3. Product Description: Strip HTML, trim whitespace

In [12]:
!pip install beautifulsoup4

In [19]:
from bs4 import BeautifulSoup

def clean_description(s):
    if pd.isnull(s):
        return s
    return BeautifulSoup(str(s), "html.parser").get_text().strip()

df['Product Description'] = df['Product Description'].apply(clean_description)

### 4. Quantity: Convert words like "two" or "3 pcs" into numeric

In [20]:
words_to_num = {
    "zero": 0, "one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
    "six": 6, "seven": 7, "eight": 8, "nine": 9, "ten": 10
}

def clean_quantity(val):
    if pd.isnull(val):
        return 0
    val_str = str(val).lower().strip()
    # Remove text like 'pcs'
    val_str = re.sub(r'[^a-zA-Z0-9]', ' ', val_str)
    for word, num in words_to_num.items():
        if word in val_str:
            return num
    nums = re.findall(r'\d+', val_str)
    return int(nums[0]) if nums else 0

df['Quantity'] = df['Quantity'].apply(clean_quantity)

### 5. Delivery Status: Normalize to "Delivered", "In Transit", "Cancelled"

In [21]:
def normalize_status(s):
    s = str(s).lower()
    if 'deliver' in s:
        return 'Delivered'
    elif 'in transit' in s or 'transit' in s or 'shipped' in s:
        return 'In Transit'
    elif 'cancel' in s:
        return 'Cancelled'
    else:
        return 'Other'

df['Delivery Status'] = df['Delivery Status'].apply(normalize_status)

## Step D: Validate Logic

### 1. Flag entries where Delivery Date < Order Date

In [22]:
df['Delivery Date'] = pd.to_datetime(df['Delivery Date'], errors='coerce')
df['Order Date'] = pd.to_datetime(df['Order Date'], errors='coerce')

df['Delivery_Date_Error'] = df['Delivery Date'] < df['Order Date']

### 2. Remove rows where Quantity <= 0 or Unit Price < 0

In [24]:
df['Quantity'] = pd.to_numeric(df['Quantity'], errors='coerce')
df['Unit Price'] = pd.to_numeric(df['Unit Price'], errors='coerce')

In [ ]:
df = df[(df['Quantity'] > 0) & (df['Unit Price'] >= 0)]

### 3. Validate that recalculated Total Price matches stored value ±1%

In [30]:
df['Quantity'] = pd.to_numeric(df['Quantity'], errors='coerce')
df['Unit Price'] = pd.to_numeric(df['Unit Price'], errors='coerce')
df['Shipping Fee'] = pd.to_numeric(df['Shipping Fee'], errors='coerce')
df['Total Price'] = pd.to_numeric(df['Total Price'], errors='coerce')

# If not already: recalculate total price
df['Recalc Total Price'] = df['Quantity'] * df['Unit Price'] + df['Shipping Fee']

# Flag rows where difference is more than 1%
df['Total_Price_Mismatch'] = (
    abs(df['Total Price'] - df['Recalc Total Price']) > 0.01 * df['Recalc Total Price']
)

In [31]:
df.head()
df.info()
df[['Delivery_Date_Error', 'Total_Price_Mismatch']].value_counts()

<class 'pandas.core.frame.DataFrame'>
Index: 2047 entries, 1 to 9998
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Order ID              1818 non-null   object        
 1   Customer Name         2047 non-null   object        
 2   Email                 2047 non-null   object        
 3   Country               2047 non-null   object        
 4   Currency              2047 non-null   object        
 5   Product SKU           2047 non-null   object        
 6   Product Description   2047 non-null   object        
 7   Quantity              2047 non-null   int64         
 8   Unit Price            2047 non-null   float64       
 9   Total Price           0 non-null      float64       
 10  Order Date            2047 non-null   datetime64[ns]
 11  Shipping Fee          1040 non-null   float64       
 12  Delivery Status       2047 non-null   object        
 13  Delivery Date         2

Delivery_Date_Error  Total_Price_Mismatch
False                False                   1511
True                 False                    536
Name: count, dtype: int64

In [32]:
df['Delivery_Date_Error'] = df['Delivery Date'] < df['Order Date']

In [33]:
df = df[(df['Quantity'] > 0) & (df['Unit Price'] >= 0)]

In [34]:
df['Recalc Total Price'] = df['Quantity'] * df['Unit Price'] + df['Shipping Fee']
df['Total_Price_Mismatch'] = (
    abs(df['Total Price'] - df['Recalc Total Price']) > 0.01 * df['Recalc Total Price']
)

In [35]:
from datetime import datetime

def get_validity(row):
    if row['Order Date'] > pd.Timestamp(datetime.now()):
        return 'Invalid: Future Date'
    if row.get('Delivery_Date_Error', False):
        return 'Invalid: Delivery Date Error'
    if row.get('Total_Price_Mismatch', False):
        return 'Invalid: Wrong Total'
    return 'Valid'

df['Order Validity'] = df.apply(get_validity, axis=1)

In [36]:
df_clean = df[df['Order Validity'] == 'Valid']
df_clean.to_csv('ecommerce_orders_cleaned.csv', index=False)

In [37]:
df.to_csv('ecommerce_orders_with_validity.csv', index=False)

In [38]:
num_valid = (df['Order Validity'] == 'Valid').sum()
num_future_dates = (df['Order Validity'] == 'Invalid: Future Date').sum()
num_wrong_totals = (df['Order Validity'] == 'Invalid: Wrong Total').sum()
num_delivery_date_errors = (df['Order Validity'] == 'Invalid: Delivery Date Error').sum()
num_removed = len(df) - num_valid

##### 3. Markdown summary

# Ecommerce Dataset Cleaning Summary

## Total Records Processed
- Original rows: **10,000**
- Final valid rows: **{num_valid}**
- Invalid records removed: **{num_removed}**

## Invalid Records Breakdown
- Future Dates: **{num_future_dates}**
- Wrong Totals: **{num_wrong_totals}**
- Delivery Date Errors: **{num_delivery_date_errors}**

## Biggest Cleanup Challenges
- **Multi-format fields:** Converting mixed currencies and price/quantity formats to consistent numeric values.
- **Missing and inconsistent data:** Many records had missing or nonsensical values for `Order ID`, dates, or prices.
- **Logical inconsistencies:** Orders with delivery dates before order dates, or impossible total prices.
- **Text normalization:** Cleaning customer names, removing emojis, and standardizing email and SKU fields.

